src/cpp/server/server_context.cc

```diff
@@ -397,10 +538,18 @@ void ServerContextBase::SetLoadReportingCosts(
   }
 }
 
-void ServerContextBase::CreateCallMetricRecorder() {
-  GPR_ASSERT(call_metric_recorder_ == nullptr);
-  grpc_core::Arena* arena = grpc_call_get_arena(call_.call);
-  call_metric_recorder_ = arena->New<experimental::CallMetricRecorder>(arena);
+void ServerContextBase::CreateCallMetricRecorder(
+    grpc_core::ServerMetricRecorder* server_metric_recorder) {
+  if (call_metric_recorder_ == nullptr) {
+    grpc_core::Arena* arena = grpc_call_get_arena(call_.call);
+    auto* backend_metric_state =
+        arena->New<BackendMetricState>(server_metric_recorder);
+    call_metric_recorder_ = backend_metric_state;
+    grpc_call_context_set(
+        call_.call, GRPC_CONTEXT_BACKEND_METRICS,
+        static_cast<grpc_core::BackendMetricProvider*>(backend_metric_state),
```

Comment:
There's no need for this cast.  The `grpc_call_context_set()` method takes a `void*` parameter, so you can pass any pointer type.

Reply:
We need this because BackendMetricProvider is the second parent. Without an explicit cast we end up storing a void* that points to the offset for BackendMetricState and CallMetricRecorder but later when the filter casts back from void* it has no type info and end up giving a corrupt BackendMetricProvider.

We can make BackendMetricProvider the first parent and remove static_cast() but I think it's better to have static_cast regardless in this case for clarity.

Reply:
Sure, done.